docs/observability/bigquery-agent-analytics.md

```diff
@@ -0,0 +1,1105 @@
+  part.object_ref.uri AS gcs_uri,
+  -- Generate a signed URL to read the content directly (requires connection_id configuration)
+  STRING(OBJ.GET_ACCESS_URL(part.object_ref, 'r').access_urls.read_url) AS signed_url
+FROM `your-gcp-project-id.your-dataset-id.agent_events_v2`,
+UNNEST(content_parts) AS part
+WHERE part.storage_mode = 'GCS_REFERENCE'
+ORDER BY timestamp DESC
+LIMIT 10;
+```
+
+## Advanced analysis queries
+
+**Trace a specific conversation turn using trace_id**
+
+```sql
+SELECT timestamp, event_type, agent, JSON_VALUE(content, '$.response') as summary
+FROM `your-gcp-project-id.your-dataset-id.agent_events_v2`
+WHERE trace_id = 'your-trace-id'
+ORDER BY timestamp ASC;
+```
+
+**Token usage analysis (accessing JSON fields)**
+
+```sql
+SELECT
+  AVG(CAST(JSON_VALUE(content, '$.usage.total') AS INT64)) as avg_tokens
+FROM `your-gcp-project-id.your-dataset-id.agent_events_v2`
+WHERE event_type = 'LLM_RESPONSE';
+```
+
+**Querying Multimodal Content (using content_parts and ObjectRef)**
+
+```sql
+SELECT
+  timestamp,
+  part.mime_type,
+  part.object_ref.uri as gcs_uri
+FROM `your-gcp-project-id.your-dataset-id.agent_events_v2`,
+UNNEST(content_parts) as part
+WHERE part.mime_type LIKE 'image/%'
+ORDER BY timestamp DESC;
+```
+
+**Analyze Multimodal Content with BigQuery Remote Model (Gemini)**
+
+```sql
+SELECT
+  logs.session_id,
+  -- Get a signed URL for the image
+  STRING(OBJ.GET_ACCESS_URL(parts.object_ref, "r").access_urls.read_url) as signed_url,
+  -- Analyze the image using a remote model (e.g., gemini-pro-vision)
+  AI.GENERATE(
+    ('Describe this image briefly. What company logo?', parts.object_ref)
+  ) AS generated_result
+FROM
+  `your-gcp-project-id.your-dataset-id.agent_events_v2` logs,
+  UNNEST(logs.content_parts) AS parts
+WHERE
+  parts.mime_type LIKE 'image/%'
+ORDER BY logs.timestamp DESC
+LIMIT 1;
+```
+
+**Latency Analysis (LLM & Tools)**
+
+```sql
+SELECT
+  event_type,
+  AVG(CAST(JSON_VALUE(latency_ms, '$.total_ms') AS INT64)) as avg_latency_ms
+FROM `your-gcp-project-id.your-dataset-id.agent_events_v2`
+WHERE event_type IN ('LLM_RESPONSE', 'TOOL_COMPLETED')
+GROUP BY event_type;
+```
+
+**Span Hierarchy & Duration Analysis**
+
+```sql
+SELECT
+  span_id,
+  parent_span_id,
+  event_type,
+  timestamp,
+  -- Extract duration from latency_ms for completed operations
+  CAST(JSON_VALUE(latency_ms, '$.total_ms') AS INT64) as duration_ms,
+  -- Identify the specific tool or operation
+  COALESCE(
+    JSON_VALUE(content, '$.tool'), 
+    'LLM_CALL'
+  ) as operation
+FROM `your-gcp-project-id.your-dataset-id.agent_events_v2`
+WHERE trace_id = 'your-trace-id'
+  AND event_type IN ('LLM_RESPONSE', 'TOOL_COMPLETED')
+ORDER BY timestamp ASC;
+```
+
+## Additional resources
+
+-   [BigQuery Storage Write API](https://cloud.google.com/bigquery/docs/write-api)
+-   [Introduction to Object Tables](https://cloud.google.com/bigquery/docs/object-tables-intro)
+
+<br>
+<br>
+<br>
+<hr>
+
+## Deprecated (v < 1.21.0)
```

Comment:
I don't think we want to include the full content of the old / deprecated version, but I will defer to @joefernandez on this.